### PR TITLE
feat(platform/bitbucket): Add ability to use UUID for reviewers

### LINKED ISSUE #16571: Bitbucket Cloud reviewers should use uuid in addition to username
### What would you like Renovate to be able to do?

BitBucket [no longer exposes the username](https://developer.atlassian.com/cloud/bitbucket/bitbucket-api-changes-gdpr/) field so its harder to configure reviewers in Renovate because you have to ask each developer what their username is.  Please consider allowing `uuid`'s to be used in additional to the existing username support to be specified as reviewers because those values can be looked up from the BitBucket API.

### If you have any ideas on how this should be implemented, please tell us here.

`addReviewers` could parse the reviewer field and if it's a UUID it could map it to the `uuid` field instead of the user `username` field.
https://github.com/renovatebot/renovate/blob/6453c3833f229c8556487549c5f481b8133e8852/lib/modules/platform/bitbucket/index.ts#L593-L605

[BitBucket API reference](https://developer.atlassian.com/cloud/bitbucket/rest/api-group-pullrequests/#api-repositories-workspace-repo-slug-pullrequests-post)
```
{
    "title": "My Title",
    "source": {
        "branch": {
            "name": "my-feature-branch"
        }
    },
    "reviewers": [
        {
            "uuid": "{504c3b62-8120-4f0c-a7bc-87800b9d6f70}"
        }
    ]
}
```

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by kwelch**
## Changes

Added a switch to use `UUID` instead of `username` for Bitbucket Cloud reviewers.

## Context

- Closes #16571

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

I didn't see any docs related to this, but if there are some I will update those as well. 

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/platform/bitbucket/index.ts [hunk 1]
```diff
@@ -623,7 +623,18 @@ export async function addReviewers(
 
   const body = {
     title,
-    reviewers: reviewers.map((username: string) => ({ username })),
+    reviewers: reviewers.map((username: string) => {
+      const UUIDRegex = new RegExp(
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:627]:
```suggestion
      const UUIDRegex = regEx(
```
use our regex helper. also very bad practice to construct same regex inside loops. 😉

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/bitbucket/index.ts (hunk 1 above)
Comment (kwelch) [reply lib/modules/platform/bitbucket/index.ts:627]: Thanks! I updated and moved. I also saw that the flags are an argument, so switched that around as well.

File: lib/modules/platform/bitbucket/index.ts [hunk 2]
```diff
@@ -621,9 +621,22 @@ export async function addReviewers(
   // TODO #7154
   const { title } = (await getPr(prId))!;
 
+  const UUIDRegex = regEx(
+    /^\{[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\}$/,
+    'i'
+  );
```
Comment (viceice) [inline lib/modules/platform/bitbucket/index.ts:627]: looks like a good candidate to extract a util function. not sure if it should be in regex or string utils.

@rarkins @secustor WDYT?

Comment (JamieMagee) [reply lib/modules/platform/bitbucket/index.ts:627]: Maybe even to [a UUID type][1] too :upside_down_face:

/s

[1]: https://ybogomolov.me/type-level-uuid/

Comment (secustor) [reply lib/modules/platform/bitbucket/index.ts:627]: Do you use somewhere else UUIDs with additional functionality? If yes, we could create a own utils module else, I would put it with the regex utils.

Comment (kwelch) [reply lib/modules/platform/bitbucket/index.ts]: I checked around and didn't find any other UUID parsing. I put it in the regex util.

File: lib/modules/platform/bitbucket/index.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import * as git from '../../../util/git';
 import * as hostRules from '../../../util/host-rules';
 import { BitbucketHttp, setBaseUrl } from '../../../util/http/bitbucket';
 import type { HttpOptions } from '../../../util/http/types';
-import { regEx } from '../../../util/regex';
+import { checkUUID, regEx } from '../../../util/regex';
```
Comment (secustor) [inline lib/modules/platform/bitbucket/index.ts:11]:
```suggestion
import { isUUID, regEx } from '../../../util/regex';
```

File: lib/modules/platform/bitbucket/index.ts [hunk 4]
```diff
@@ -623,7 +623,15 @@ export async function addReviewers(
 
   const body = {
     title,
-    reviewers: reviewers.map((username: string) => ({ username })),
+    reviewers: reviewers.map((username: string) => {
+      let key = 'username';
+      if (checkUUID(username)) {
+        key = 'uuid';
+      }
```
Comment (secustor) [inline lib/modules/platform/bitbucket/index.ts:630]:
```suggestion
      const key = isUUID(username) ? 'uuid' : 'username';
```
small simplification

File: lib/util/regex.ts [hunk 5]
```diff
@@ -98,3 +98,12 @@ export function configRegexPredicate(
   }
   return null;
 }
+
+const UUIDRegex = regEx(
+  /^\{[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\}$/,
+  'i'
+);
+
+export function checkUUID(input: string): boolean {
```
Comment (secustor) [inline lib/util/regex.ts:107]:
```suggestion
export function isUUID(input: string): boolean {
```
`is` communicates that we will return a boolean.

**Comment by kwelch** [comment]
Switched to better name with `is`, and simplified the call.

**Review by secustor — DISMISSED** [review]

**Review by JamieMagee — DISMISSED** [review]

File: lib/util/regex.ts [hunk 6]
```diff
@@ -98,3 +98,12 @@ export function configRegexPredicate(
   }
   return null;
 }
+
+const UUIDRegex = regEx(
+  /^\{[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}\}$/,
+  'i'
+);
+
+export function isUUID(input: string): boolean {
```
Comment (viceice) [inline lib/util/regex.ts:107]: please add a test to `./regex.spec.ts`, so this function is covered when only running tests inside the utils folder.

Comment (kwelch) [reply lib/util/regex.ts:107]: Added one, it is pretty basic but should work. I tried to match convention here as much as possible.

**Review by viceice — APPROVED** [review]

**Review by secustor — APPROVED** [review]

**Pull request merged**